src/main/java/org/casbin/jcasbin/main/SyncedEnforcer.java

```diff
@@ -95,22 +96,819 @@ public SyncedEnforcer(String modelPath, String policyFile, boolean enableLog) {
         super(modelPath, policyFile, enableLog);
     }
 
+    /**
+     * setWatcher sets the current watcher.
+     *
+     * @param watcher the watcher.
+     */
+    public void setWatcher(Watcher watcher) {
+        this.watcher = watcher;
+        watcher.setUpdateCallback(this::loadPolicy);
+    }
+
+    /**
+     * clearPolicy clears all policy.
+     */
+    public void clearPolicy() {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            super.clearPolicy();
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * loadPolicy reloads the policy from file/database.
+     */
+    public void loadPolicy() {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            super.loadPolicy();
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * loadFilteredPolicy reloads a filtered policy from file/database.
+     *
+     * @param filter the filter used to specify which type of policy should be loaded.
+     */
+    public void loadFilteredPolicy(Object filter) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            super.loadFilteredPolicy(filter);
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * savePolicy saves the current policy (usually after changed with
+     * Casbin API) back to file/database.
+     */
+    public void savePolicy() {
+        try {
+            READ_WRITE_LOCK.readLock().lock();
+            super.savePolicy();
+        } finally {
+            READ_WRITE_LOCK.readLock().unlock();
+        }
+    }
+
+    /**
+     * buildRoleLinks manually rebuild the
+     * role inheritance relations.
+     */
+    public void buildRoleLinks() {
+        try {
+            READ_WRITE_LOCK.readLock().lock();
```

Comment:
Why not use `writeLock`?  @hsluoyz , same question in Golang.

Reply:
Is Golang wrong here?

Reply:
I think it is better, we use readLock here, as it does not involve any manipulation of config data, it just adds role links to the RoleManager and should not be concurrent with any of the getters functions.

Reply:
`RoleManager` need read `Model`.

Reply:
It cleans up the `RoleManager`.